src/src/PluginStructs/P145_data_struct.cpp

```diff
@@ -0,0 +1,598 @@
+    {
+      sum -= (float)ovs_max;                   // remove OversamplingMaxVal
+      sum -= (float)ovs_min;                   // remove OversamplingMinVal
+      count -= 2;
+    }
+    ain = sum / (float)count;                  // Scale to single sample value
+    last_ain = (unsigned long)(sum / 100);     // Remember result
+  }
+  return ain;
+}
+
+/*****************************************************************************/
+/*!
+@brief  Set the configuration data for the sensor  
+@param[in] stype   Sensor type [enum]
+@param[in] comp    Temperature/humidity compensation enable/disable
+@param[in] cal     Autocalibration enable/disable
+@param[in] load    Load resistence Rload
+@param[in] zero    Rzero, Sensor resistor value for reference level
+@param[in] ref     Reference level for calibration
+@note   These parameters must be set before the plugin can calculate the level
+        They are determined by the plugin configuration
+/*****************************************************************************/
+void P145_data_struct::setSensorData(int stype, bool comp, bool cal, float load, float zero, float ref)
+{
+  /* Each MQ-xxx sensor comes with its own set cof constants */
+  /* Copy the correct set from program meory space           */
+  if ((stype != sensorType) && (stype < nbrOfTypes) && (stype >= 0))
+  {
+    memcpy_P(&sensordef, &sensorData[stype], sizeof(struct P145_SENSORDEF));
+    algorithm = sensordef.alg;
+    sensorType = stype;   // Selected sensor type, index in sensor data table
+  }
+  compensation = comp;    // Compensation selection flag
+  calibration = cal;      // Calibration selection flag
+  rload = load;           // Rload, load resistor [Ohm]
+  rzero = zero;           // R0, reference resistance [Ohm]
+  refLevel = ref;         // Reference level for calibration [ppm]
+}
+
+/*****************************************************************************/
+/*!
+@brief  Return the current/latest value acquired from the sensor 
+@param[in] temperature  Temperature for compensation
+@param[in] humidity     Humidity for compensation
+@return The calculated gas concentration [ppm or mg/l]
+@note   This is the entry point to start collecting the oversampling data,
+        convert the analog value to Rsensor and convert Rsensor to the gas
+        concentration level. It also performs autocalibration if selected.
+        Oversampling data will be reset as a side effec.
+        This function shall be called at regular interval for proper performance.
+*/
+/*****************************************************************************/
+float P145_data_struct::readValue(float temperature, float humidity)
+{
+    float ain = getAnalogValue();           // Analog measured and filtered data 
+    float rSensor = getResistance(ain);     // Convert to sensor resistance Rs
+    float rCal    = 0.0;                    // Potential Rzero for calibration
+    float value = 0.0;                      // Return value
```

Comment:
`0.0` instead of `0.0f`
This means the compiler will make a double type first and then casts it to a float.
There are several other places in this code where the `f` is missing.